dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -973,15 +1150,383 @@ class FunctionStringRegexp : public FunctionStringRegexpBase
+{
+public:
+    using ResultType = Int64;
+    static constexpr auto name = Name::name;
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionStringRegexpInstr>(); }
+    String getName() const override { return name; }
+    bool isVariadic() const override { return true; }
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        size_t arg_num = arguments.size();
+        if (arg_num < REGEXP_MIN_PARAM_NUM)
+            throw Exception("Too few arguments", ErrorCodes::TOO_LESS_ARGUMENTS_FOR_FUNCTION);
+        else if (arg_num > REGEXP_INSTR_MAX_PARAM_NUM)
+            throw Exception("Too many arguments", ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION);
+
+        bool has_nullable_col = false;
+        bool has_data_type_nothing = false;
+        bool is_str_arg;
+
+        // Check type of arguments
+        for (size_t i = 0; i < arg_num; ++i)
+        {
+            // Index at 0, 1 and 5 arguments should be string type, otherwise int type.
+            is_str_arg = (i <= 1 || i == 5);
+            checkInputArg(arguments[i], is_str_arg, &has_nullable_col, &has_data_type_nothing);
+        }
+
+        if (has_data_type_nothing)
+            return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNothing>());
+
+        if (has_nullable_col)
+            return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNumber<ResultType>>());
+        else
+            return std::make_shared<DataTypeNumber<ResultType>>();
+    }
+
+    template <typename ExprT, typename PatT, typename PosT, typename OccurT, typename RetOpT, typename MatchTypeT>
+    void REGEXP_CLASS_MEM_FUNC_IMPL_NAME(ColumnWithTypeAndName & res_arg, const ExprT & expr_param, const PatT & pat_param, const PosT & pos_param, const OccurT & occur_param, const RetOpT & ret_op_param, const MatchTypeT & match_type_param) const
+    {
+        size_t col_size = expr_param.getDataNum();
+
+        // Get function pointers to process the specific int type
+        GetIntFuncPointerType get_pos_func = getGetIntFuncPointer(pos_param.getIntType());
+        GetIntFuncPointerType get_occur_func = getGetIntFuncPointer(occur_param.getIntType());
+        GetIntFuncPointerType get_ret_op_func = getGetIntFuncPointer(ret_op_param.getIntType());
+
+        // Container will not be used when parm is const
+        const void * pos_container = pos_param.getContainer();
+        const void * occur_container = occur_param.getContainer();
+        const void * ret_op_container = ret_op_param.getContainer();
+
+        // Const value will not be used when param is not const
+        Int64 pos_const_val = PosT::isConst() ? pos_param.template getInt<Int64>(0) : -1;
```

Comment:
> What happens if `col_size == 0`? Why not just use `Int64 pos_const_val = -1` since we already make sure that `pos_const_val` is only used when `PosT::isConst()` is true



> Also need to add some fullstack tests. And since regexp related function have different behavior if the input is empty or not, we need add test of empty input explicitly.

okk

Reply:
> What happens if `col_size == 0`? Why not just use `Int64 pos_const_val = -1` since we already make sure that `pos_const_val` is only used when `PosT::isConst()` is true

nothing will happen when `col_size == 0` as the index parameter passed into `getInt<Int64>` is useless when this is a const value. You can see the implementation of `getInt` in `ParamInt`.